svcdef/svcdef.go

```diff
@@ -0,0 +1,354 @@
+		fset := token.NewFileSet()
+		fileAst, err := parser.ParseFile(fset, "", gofile, parser.ParseComments)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't parse go file to create Svcdef")
+		}
+
+		typespecs, _ := retrieveTypeSpecs(fileAst)
+		for _, t := range typespecs {
+			//sp.Dump(t)
+			switch typdf := t.Type.(type) {
+			case *ast.Ident:
+				if typdf.Name == "int32" {
+					nenm, err := NewEnum(t)
+					if err != nil {
+						return nil, errors.Wrapf(err, "error parsing enum %q", t.Name.Name)
+					}
+					rv.Enums = append(rv.Enums, nenm)
+				}
+			case *ast.StructType:
+				nmsg, err := NewMessage(t)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error parsing message %q", t.Name.Name)
+				}
+				rv.Messages = append(rv.Messages, nmsg)
+			case *ast.InterfaceType:
+				// Each service will have two interfaces ("{SVCNAME}Server" and
+				// "{SVCNAME}Client") each containing the same information that we
+				// care about, but structured a bit differently. For simplicity,
+				// skip the "Client" interface.
+				if strings.HasSuffix("Client", t.Name.Name) {
+					break
+				}
+				nsvc, err := NewService(t)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error parsing service %q", t.Name.Name)
+				}
+				rv.Service = nsvc
+			}
+		}
+	}
+	resolveTypes(&rv)
+	err := consolidateHTTP(&rv, protoFiles)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to consolidate HTTP")
+	}
+
+	return &rv, nil
+}
+
+func NewEnum(e *ast.TypeSpec) (*Enum, error) {
+	return &Enum{
+		Name: e.Name.Name,
+	}, nil
+}
+
+// NewMessage returns a new Message struct derived from an *ast.TypeSpec with a
+// Type of *ast.StructType.
+func NewMessage(m *ast.TypeSpec) (*Message, error) {
+	rv := &Message{
+		Name: m.Name.Name,
+	}
+
+	strct := m.Type.(*ast.StructType)
+	for _, f := range strct.Fields.List {
+		nfield, err := NewField(f)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't create field %q while creating message %q", f.Names[0].Name, rv.Name)
+		}
+		rv.Fields = append(rv.Fields, nfield)
+	}
+
+	return rv, nil
+}
+
+// NewMap returns a new Map struct derived from an ast.Expr interface
+// implemented by an *ast.MapType struct. This code cannot accept an arbitrary
+// MapType, only one which follows the conventions of Go code generated by
+// protoc-gen-go. Those conventions are:
+//
+//     1. The KeyType of the *ast.MapType will always be an ast.Ident
+//     2. The ValueType may be an ast.Ident OR an ast.StarExpr -> ast.Ident
+//
+// These rules are a result of the rules for map fields of Protobuf messages,
+// namely that a key may only be represented by a non-float basetype (e.g.
+// int64, string, etc.), and that a value may be either a basetype or a Message
+// type or an Enum type. In the resulting Go code, a basetype will be
+// represented as an ast.Ident, while a key that is a Message or Enum type will
+// be represented as an *ast.StarExpr which references an ast.Ident.
```

Comment:
> //This code cannot accept an arbitrary
> // MapType, only one which follows the conventions of Go code generated by
> // protoc-gen-go.

It might be a good idea to add a package level description of svcdef saying

```
// package svcdef provides types used to represent a grpc Service with protobuf messages. 
// It implements a parser for Go code generated by protoc-gen-go and all methods accept 
// only ast of the form created by protoc-gen-go.
```

Then include these specifics within the function body to be read by someone working on the function itself.